lib/router.ex

```diff
@@ -8,12 +9,12 @@ defmodule HTTPServer.Router do
   @routes Application.compile_env(:http_server, :routes, Routes)
 
   def router(req) do
-    case Map.fetch(@routes.routes, req.path) do
-      {:ok, path_info} ->
-        router(req, path_info)
-
-      _ ->
-        route_not_found(req)
+    with {:ok, path_info} <- Map.fetch(@routes.routes, req.path),
+         true <- is_method_allowed(req.method, path_info) do
+      router(req, path_info)
+    else
+      :error -> route_not_found(req)
+      false -> method_not_allowed(req)
```

Comment:
```Elixir
    else
      :error -> route_not_found(req)
      false -> method_not_allowed(req)
```
is really cool, and I understand what's happening here.

```Elixir
    with {:ok, path_info} <- Map.fetch(@routes.routes, req.path),
         true <- is_method_allowed(req.method, path_info) do
      router(req, path_info)
```
This makes me a bit dizzy, and its not obvious that there is something about the `is_method_allowed/2` that gets forwarded to the `else` clause.

Is there a way to make this a bit more understandable for our future (and my current) selves?

Reply:
I could create a boolean `is_route_not_found/1`, rework my methods boolean to be `is_method_not_allowed/2`, then change this to be a conditional statement:
```
cond do
  is_route_not_found(req.path) ->
    route_not_found(req)

  is_method_not_allowed(req.method, path_info) ->
    method_not_allowed(req)

  true ->
    router(req, path_info)
end
```

Reply:
It is an improvement, because its more readable.

Maybe some extraction of responsibilities could be as well.

```suggestion
    case Map.fetch(@routes.routes, req.path) do
      {:ok, path_info} -> route_if_allowed(req, path_info)
      _ -> route_not_found(req)
    end

  defp route_if_allowed(req = %Request{method: method}, path_info) do
    cond do
      !is_method_allowed(method, path_info) -> method_not_allowed(req)
      true -> router(req, path_info)
    end
  end
```

Reply:
all three suggestions are correct.  Some are more readable than others, and there may be yet more. So, I am good with whatever you eventually choose.

Reply:
Love the extraction of responsibility! Just pushed up this change, but instead of a `cond` block in `route_if_allowed/2` I used case:
```
  defp route_if_allowed(req = %Request{method: method}, path_info) do
    case is_method_allowed(method, path_info) do
      true -> router(req, path_info)
      false -> method_not_allowed(req)
    end
  end
```

Reply:
nice